src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs

```diff
@@ -45,74 +43,62 @@ internal sealed class InlineHintsTagger : ITagger<IntraTextAdornmentTag>, IDispo
 
         private readonly InlineHintsTaggerProvider _taggerProvider;
 
-        private readonly ITextBuffer _buffer;
         private readonly IWpfTextView _textView;
-
-        public event EventHandler<SnapshotSpanEventArgs>? TagsChanged;
+        private readonly ITextBuffer _subjectBuffer;
 
         public InlineHintsTagger(
             InlineHintsTaggerProvider taggerProvider,
             IWpfTextView textView,
-            ITextBuffer buffer,
-            ITagAggregator<InlineHintDataTag> tagAggregator)
+            ITextBuffer subjectBuffer,
+            EfficientTagger<InlineHintDataTag> tagger)
         {
             _taggerProvider = taggerProvider;
 
             _textView = textView;
-            _buffer = buffer;
+            _subjectBuffer = subjectBuffer;
+
+            // When the underlying tagger produced new data tags, inform any clients of us that we have new adornment tags.
+            _underlyingTagger = tagger;
+            _underlyingTagger.TagsChanged += OnTagsChanged;
 
-            _tagAggregator = tagAggregator;
             _formatMap = taggerProvider.ClassificationFormatMapService.GetClassificationFormatMap(textView);
             _hintClassification = taggerProvider.ClassificationTypeRegistryService.GetClassificationType(InlineHintsTag.TagId);
+
             _formatMap.ClassificationFormatMappingChanged += this.OnClassificationFormatMappingChanged;
-            _tagAggregator.BatchedTagsChanged += TagAggregator_BatchedTagsChanged;
+            _taggerProvider.GlobalOptionService.AddOptionChangedHandler(this, OnGlobalOptionChanged);
         }
 
-        /// <summary>
-        /// Goes through all the spans in which tags have changed and
-        /// invokes a TagsChanged event. Using the BatchedTagsChangedEvent since it is raised
-        /// on the same thread that created the tag aggregator, unlike TagsChanged.
-        /// </summary>
-        private void TagAggregator_BatchedTagsChanged(object sender, BatchedTagsChangedEventArgs e)
+        public override void Dispose()
         {
-            _taggerProvider.ThreadingContext.ThrowIfNotOnUIThread();
-            InvalidateCache();
```

Comment:
no need for invalidation when the underlying tagger produces tags.  adornment tags are directly associated with the data tags, allowing them to be GC'ed when the data tag goes away.